strawberryfields/apps/graph/dense.py

```diff
@@ -0,0 +1,179 @@
+
+.. math:: d(G[S]) = \frac{2 |E_{S}|}{|S|(|S|-1)},
+
+where :math:`|\cdot|` denotes cardinality, and the densest-:math:`k` subgraph problem can be
+written succinctly as
+
+.. math:: {\rm argmax}_{S \in \mathcal{S}_{k}} d(G[S])
+
+with :math:`\mathcal{S}_{k}` the set of all possible :math:`k` node subgraphs. This problem grows
+combinatorially with :math:`{N \choose k}` and is NP-hard in the worst case.
+
+The :func:`find_dense` function provides access to heuristic algorithms for finding
+approximate solutions. At present, random search is the heuristic algorithm provided, accessible
+through the :func:`random_search` function. This algorithm proceeds by randomly generating a set
+of :math:`k` vertex subgraphs and selecting the densest. Sampling of subgraphs can be achieved
+both uniformly at random and also with a quantum sampler programmed to be biased toward
+outputting dense subgraphs.
+
+Summary
+-------
+
+.. autosummary::
+    METHOD_DICT
+    OPTIONS_DEFAULTS
+    find_dense
+    random_search
+
+Code details
+------------
+"""
+from typing import Tuple, Optional
+import networkx as nx
+from ..sample import BACKEND_DEFAULTS
+from . import resize, sample, utils
+from .utils import graph_type
+
+
+def find_dense(
+    graph: graph_type, nodes: int, iterations: int = 1, options: Optional[dict] = None
+) -> Tuple[float, list]:
+    """Find a dense subgraph of a given size.
+
+    This function returns the densest `node-induced subgraph
+    <http://mathworld.wolfram.com/Vertex-InducedSubgraph.html>`__ of size ``nodes`` after
+    multiple repetitions. It uses heuristic optimization that combines search space exploration
+    with local searching. The heuristic method can be set with the ``options`` argument. Methods
+    can contain stochastic elements, where randomness can come from distributions including GBS.
+
+    All elements of the heuristic can be controlled with the ``options`` argument, which should be a
+    dict-of-dicts where the first level specifies the option type as a string-based key,
+    with corresponding value being a dictionary of options for that type. The option types are:
+
+    - ``"heuristic"``: specifying options used by optimization heuristic; corresponding
+      dictionary of options explained further :ref:`below <heuristic>`
+    - ``"backend"``: specifying options used by backend quantum samplers; corresponding
+      dictionary of options explained further in :mod:`~strawberryfields.future.apps.sample`
+    - ``"resize"``: specifying options used by resizing method; corresponding dictionary of
+      options explained further in :mod:`~strawberryfields.future.apps.graph.resize`
+    - ``"sample"``: specifying options used in sampling; corresponding dictionary of options
+      explained further in :mod:`~strawberryfields.future.apps.graph.sample`
+
+    If unspecified, a default set of options is adopted for a given option type.
+
+    .. _heuristic:
```

Comment:
Does this work? I thought that Sphinx references only worked above section titles